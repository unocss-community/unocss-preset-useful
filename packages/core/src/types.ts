import type { CSSObject, Preset, SourceCodeTransformer, StaticShortcut } from '@unocss/core'
// Presets
import type { AttributifyOptions } from '@unocss/preset-attributify'
import type { IconsOptions } from '@unocss/preset-icons'
import type { RemToPxOptions } from '@unocss/preset-rem-to-px'
import type { TagifyOptions } from '@unocss/preset-tagify'
import type { TypographyOptions } from '@unocss/preset-typography'
import type { WebFontsOptions } from '@unocss/preset-web-fonts'
import type { PresetWind3Options, Theme as ThemeWind3 } from '@unocss/preset-wind3'
import type { PresetWind4Options, Theme as ThemeWind4 } from '@unocss/preset-wind4'
// Transformers
import type { CompileClassOptions } from '@unocss/transformer-compile-class'
import type { TransformerDirectivesOptions } from '@unocss/transformer-directives'
import type { TransformerVariantGroupOptions } from '@unocss/transformer-variant-group'

type CustomStaticShortcut = [string | string[], StaticShortcut[1]] | [string | string[], StaticShortcut[1], StaticShortcut[2]]
export type CustomStaticShortcuts = CustomStaticShortcut[]

export type Objectiable<T> = Record<string, T>

export type CSSKeyframesRule = Objectiable<CSSObject>

export interface UsefulExtends extends Exclude<UsefulTheme, 'extend'> {
  keyframes?: Record<string, CSSKeyframesRule>
  /**
   * Different from the original, you can use the following formats:
   *
   * ```ts
   * { name : 'name duration timing-function iteration-count' }
   * ```
   */
  animation?: Objectiable<string>
}

export interface UsefulTheme extends Omit<ThemeWind3, 'container' | 'containers'>, ThemeWind4 {
  extend?: UsefulExtends
}

export interface PostprocessOptions {
  /**
   * Make all unitilities important.
   *
   * @default false
   */
  important?: boolean | ImportantOptions

  /**
   * Extract rgba color in css variable, default key name is `--un-color`
   *
   * Only works when `wind3` is enabled
   *
   * @default false
   */
  unColor?: boolean | string
}

export type FilterPattern = Array<string | RegExp> | string | RegExp | null

export interface ImportantOptions {
  /**
   * Make all unitilities important.
   *
   */
  includes?: FilterPattern

  /**
   * Make all unitilities important.
   *
   */
  excludes?: FilterPattern
}

export interface UsefulOptions {
  /**
   * Enable built-in postprocess
   *
   * @default false
   */
  postprocess?: boolean | PostprocessOptions

  shortcuts?: boolean | {
    /**
     * Enable default shortcuts
     *
     * @default true
     */
    default?: boolean
  }

  /**
   * Improve theme to be more useful, and align with Tailwind theme configuration
   *
   * - Add `animation` to theme, Expand theme animation name usage
   *
   * [ name, duration, timing-function, iteration-count ]
   *
   * @example
   *
   * ```ts
   * theme: {
   *   extend: {
   *     animation: {
   *      shape: 'shape 5s linear infinite'
   *     },
   *     // ...
   *   }
   * }
   * ```
   * You can choose to use special symbols as placeholders, to indicate whether to inject this property into the uno theme
   *
   * - `*` Abandon injection
   * - `+` Injection, but the value is empty
   *
   * @example
   *
   * ```ts
   * theme: {
   *   extend: {
   *     animation: {
   *      foo: 'foo 1s * 3',
   *      bar: 'bar 1s +',
   *     },
   *     // ...
   *   }
   * }
   * ```
   *
   */
  theme?: UsefulTheme

  /**
   * Enable the default preset for preset-wind3
   * Only works when `presets` is not specified
   *
   * @about [@unocss/preset-wind3](https://unocss.dev/presets/wind3)
   * @default false
   */
  wind3?: boolean | PresetWind3Options

  /**
   * Enable the default preset for preset-wind4
   * Only works when `presets` is not specified
   *
   * After v1.0.0, wind4 is the default preset
   *
   * @about [@unocss/preset-wind4](https://unocss.dev/presets/wind4)
   * @default true
   */
  wind4?: boolean | PresetWind4Options

  /**
   * Enable attributify mode and the options of it
   * Only works when `presets` is not specified
   *
   * @about [@unocss/preset-attributify](https://unocss.dev/presets/attributify)
   * @default false
   */
  attributify?: boolean | AttributifyOptions

  /**
   * Enable icons preset and the options of it
   * Only works when `presets` is not specified
   *
   * @about [@unocss/preset-icons](https://unocss.dev/presets/icons)
   * @default false
   */
  icons?: boolean | IconsOptions

  /**
   * Enable webFonts preset and the options of it
   * Only works when `presets` is not specified
   *
   * **Note:** Default by [`fontsource`](https://fontsource.org/) provider
   *
   * @about [@unocss/preset-web-fonts](https://unocss.dev/presets/web-fonts)
   * @default false
   */
  webFonts?: boolean | WebFontsOptions

  /**
   * Enable typography preset and the options of it
   * Only works when `presets` is not specified
   *
   * @about [@unocss/preset-typography](https://unocss.dev/presets/typography)
   * @default false
   */
  typography?: boolean | TypographyOptions

  /**
   * Enable tagify preset and the options of it
   * Only works when `presets` is not specified
   *
   * @about [@unocss/preset-tagify](https://unocss.dev/presets/tagify)
   * @default false
   */
  tagify?: boolean | TagifyOptions

  /**
   * Enable remToPx preset and the options of it
   * Only works when `presets` is not specified
   *
   * @about [@unocss/preset-rem-to-px](https://unocss.dev/presets/rem-to-px)
   * @default false
   */
  remToPx?: boolean | RemToPxOptions

  /**
   * Enable magicss preset
   *
   * @about [unocss-preset-magicss](https://github.com/unpreset/unocss-preset-magicss)
   * @default false
   */
  magicss?: boolean

  /**
   * Enable directives transformer and the options of it
   *
   * @about [@unocss/transformer-directives](https://unocss.dev/transformers/directives)
   * @default true
   */
  directives?: boolean | TransformerDirectivesOptions

  /**
   * Enables the variant group feature of Windi CSS for UnoCSS.
   *
   * @about [@unocss/transformer-variant-group](https://unocss.dev/transformers/variant-group)
   * @default true
   */
  variantGroup?: boolean | TransformerVariantGroupOptions

  /**
   * Compile group of classes into one class
   *
   * @about [@unocss/transformer-class-group](https://unocss.dev/transformers/compile-class)
   * @default false
   */
  compileClass?: boolean | CompileClassOptions
}

export type ResolvedOptions = Required<UsefulOptions> & {
  meta: {
    presets: Preset[]
    shortcuts: CustomStaticShortcuts
    transformers: SourceCodeTransformer[]
  }
}

export type DeepPartial<T> = { [P in keyof T]?: DeepPartial<T[P]> }
