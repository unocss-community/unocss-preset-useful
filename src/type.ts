export interface UsefulOptions {
  /**
   * Extract rgba color in css variable
   *
   * @default false
   */
  unColor?: boolean | string

  /**
   * Expand theme animation name usage
   *
   * [ name, duration, timing-function, iteration-count ]
   *
   * @example
   * ```ts
    themeAnimate: ['spin 1s linear infinite'],
   * ```
   */
  themeAnimate?: string[]
}

export type ResolvedOptions = Required<UsefulOptions>

export type DeepPartial<T> = { [P in keyof T]?: DeepPartial<T[P]> }
