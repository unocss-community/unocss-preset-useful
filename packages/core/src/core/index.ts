export * from './extractors'
export * from './postprocess'
export * from './rules'
export * from './shortcuts'
export * from './theme'
