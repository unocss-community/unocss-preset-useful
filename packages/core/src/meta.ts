import type { RuleMeta } from '@unocss/core'

export const PRESET_NAME = 'useful'

export const layerMeta: RuleMeta = {
  layer: PRESET_NAME,
}
