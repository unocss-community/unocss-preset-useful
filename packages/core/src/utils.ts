import type { ThemeAnimation } from '@unocss/preset-mini'
import postcss from 'postcss'
import postcssJs, { objectify } from 'postcss-js'
import type { CSSObject } from 'unocss'
import type { CustomStaticShortcuts, DeepPartial, Objectiable } from './types'

// name duration timing-function iteration-count
export const animationRegExp = /^([a-z-]+)\s+([0-9.]+m?s?|[*+])?\s?([a-z-]+(?:\([^)]+\))?|[*+])?\s*([a-z0-9-]+|[*+])?$/i

/**
 * Normalize custom animate usage to UnoCSS animations theme.
 *
 * ⚠️ You must strictly follow the following format. ⚠️
 *
 * [ name, duration, timing-function, iteration-count ]
 *
 * If you use * as placeholder, it will be ignored.
 *
 * [name, duration, *, iteration-count]
 *
 * If you use + as placeholder, it will be replaced with empty string.
 *
 * [name, duration, +, iteration-count]
 *
 * @example
 *
 * { animate: ['spin 1s linear infinite'] }
 *
 * Will be transformd:
 *
 * {
 *   animate: ['spin 1s linear infinite'],
 *   durations: {
 *     spin: '1s',
 *   },
 *   timingFns: {
 *     spin: 'linear',
 *   },
 *   counts: {
 *     spin: 'infinite',
 *   },
 * }
 */
export function resolveAnimation(extend_animation: Objectiable<string>) {
  const animation: ThemeAnimation = {}
  const keys: (Exclude<keyof ThemeAnimation, 'properties'>)[] = ['durations', 'timingFns', 'counts']
  const shortcuts: CustomStaticShortcuts = []

  for (const k in extend_animation) {
    const v = extend_animation[k]
    const match = v.match(animationRegExp)
    if (match != null) {
      const [, name, duration, timing, count] = match
      const values = [duration, timing, count]

      if (name !== k)
        shortcuts.push([`animate-${k}`, `animate-${name}`])

      for (let i = 0; i < keys.length; i++) {
        const key = keys[i]
        const value = values[i]
        if (value != null) {
          if (value === '*') {
            continue
          }

          if (animation[key]) {
            animation[key]![name] = values[i] === '+' ? '' : values[i]
          }
          else {
            animation[key] = {
              [name]: values[i] === '+' ? '' : values[i],
            }
          }
        }
      }
    }
  }

  return {
    animation,
    shortcuts,
  }
}

/**
 * Deep merge two objects.
 * @param original Original object
 * @param patch Patch object
 * @returns Merged object
 */
export function deepMerge<T>(original: T, patch: DeepPartial<T>): T {
  const o = original as any
  const p = patch as any

  if (Array.isArray(o) && Array.isArray(p))
    return [...o, ...p] as any

  if (Array.isArray(o))
    return [...o] as any

  const output = { ...o }
  if (isObject(o) && isObject(p)) {
    Object.keys(p).forEach((key) => {
      if (isObject(p[key])) {
        if (!(key in o))
          Object.assign(output, { [key]: p[key] })
        else
          output[key] = deepMerge(o[key], p[key])
      }
      else {
        Object.assign(output, { [key]: p[key] })
      }
    })
  }
  return output
}

export function isObject(val: unknown): val is Record<any, any> {
  return val !== null && typeof val === 'object'
}

export function getKeyframes(css: string) {
  const root = postcss.parse(css)
  const obj = objectify(root as any)

  return Object.keys(obj).reduce<Record<string, Record<string, CSSObject>>>((acc, key) => {
    if (key.startsWith('@keyframes'))
      acc[key] = obj[key]
    return acc
  }, {})
}

export function cssObj2StrSync(style: Record<string, CSSObject>): string {
  return Object.keys(style).reduce((str, key) => {
    return `${str}${key}${stringifyObj(style[key])}`
  }, '').replace(/\n/g, '')
}

export function stringifyObj(obj: CSSObject) {
  return `{${Object.keys(obj).reduce((str, key) => {
    return `${str}${camelToHyphen(key)}:${obj[key]};`
  }, '')}}`
}

export async function cssObj2StrAsync(style: Record<string, CSSObject>) {
  return postcss().process(style, { parser: postcssJs } as any).then(result => result.css)
}

export function toArray<T>(val: T | T[]): T[] {
  return Array.isArray(val) ? val : [val]
}

export function camelToHyphen(str: string): string {
  return str.replace(/([a-z])([A-Z])/g, '$1-$2').toLowerCase()
}
