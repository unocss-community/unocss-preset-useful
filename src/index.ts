import { definePreset } from 'unocss'
import type { Postprocessor, Preset } from 'unocss'
import { rules } from './rules'
import { shortcuts } from './shortcuts'
import { PRESET_NAME } from './constants'
import { extractors } from './extractors'
import { postprocessWithUnColor } from './postprocess'
import { nomarlizeTheme } from './theme'

export interface UsefulOptions {
  /**
   * Extract rgba color in css variable
   *
   * @default false
   */
  unColor?: boolean | string

  /**
   * Expand theme animation name usage
   *
   * [ name, duration, timing-function, iteration-count ]
   *
   * @example
   * ```ts
    themeAnimate: ['spin 1s linear infinite'],
   * ```
   */
  themeAnimate?: string[]
}

export function presetUseful(options: UsefulOptions = {}): Preset {
  let { unColor, themeAnimate } = options
  unColor = typeof unColor === 'string'
    ? unColor
    : unColor ? '--un-color' : false

  return definePreset({
    name: `unocss-preset-${PRESET_NAME}`,
    layers: {
      [PRESET_NAME]: 2,
    },
    rules,
    theme: {
      animation: nomarlizeTheme(themeAnimate),
    },
    shortcuts,
    extractors,
    postprocess: [
      unColor ? postprocessWithUnColor(unColor as string) : undefined,
    ].filter(Boolean) as Postprocessor[],
  })
}

export default presetUseful
