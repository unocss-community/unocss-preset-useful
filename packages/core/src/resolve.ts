import type { Theme } from '@unocss/preset-mini'
import { nomarlizeTheme } from './core'
import type { CustomStaticShortcuts, ResolvedOptions, UsefulOptions, UsefulTheme } from './types'
import { cssObj2StrSync, deepMerge, resolveAnimation } from './utils'

const defaultOptions: UsefulOptions = {
  uno: true,
  attributify: true,
  icons: true,
  webFonts: false,
  typography: false,
  tagify: false,
  remToPx: false,
  scrollbar: false,
  theme: {},
  enableDefaultShortcuts: true,
  enableMagicAnimations: false,
  important: false,
  enableResetStyles: true,
}

export async function resolveOptions(options: UsefulOptions) {
  const optionsWithDefault = Object.assign({}, defaultOptions, options) as Required<UsefulOptions>
  optionsWithDefault.unColor = typeof optionsWithDefault.unColor === 'string'
    ? optionsWithDefault.unColor
    : optionsWithDefault.unColor ? '--un-color' : false

  const presets = []
  const presetMap = {
    uno: import('@unocss/preset-uno').then(({ presetUno }) => presetUno),
    attributify: import('@unocss/preset-attributify').then(({ presetAttributify }) => presetAttributify),
    icons: import('@unocss/preset-icons').then(({ presetIcons }) => presetIcons),
    webFonts: import('@unocss/preset-web-fonts'),
    typography: import('@unocss/preset-typography').then(({ presetTypography }) => presetTypography),
    tagify: import('@unocss/preset-tagify').then(({ presetTagify }) => presetTagify),
    remToPx: import('@unocss/preset-rem-to-px').then(({ presetRemToPx }) => presetRemToPx),
    scrollbar: import('unocss-preset-scrollbar').then(({ presetScrollbar }) => presetScrollbar),
  }

  for (const [key, preset] of Object.entries(presetMap)) {
    const option = optionsWithDefault[key as keyof typeof presetMap]
    if (option) {
      const p = await preset as any
      presets.push(p(typeof option === 'boolean' ? {} as any : option))
    }
  }

  const { theme: t_theme, shortcuts } = resolveExtend(optionsWithDefault.theme.extend ?? {})
  const _theme = deepMerge(optionsWithDefault.theme, t_theme)

  return {
    ...optionsWithDefault,
    theme: nomarlizeTheme(_theme, optionsWithDefault.enableMagicAnimations),
    meta: {
      presets,
      shortcuts,
    },
  } as ResolvedOptions
}

export function resolveExtend(extend: UsefulTheme['extend']) {
  const _shortcuts: CustomStaticShortcuts = []
  const { animation, keyframes } = extend!

  // animation
  const { animation: resolvedAnimation, shortcuts } = resolveAnimation(animation ?? {})
  _shortcuts.push(...shortcuts)

  // keyframes
  resolvedAnimation.keyframes = {}
  for (const key in keyframes)
    resolvedAnimation.keyframes[key] = `{${cssObj2StrSync(keyframes[key])}}`

  return {
    theme: { animation: resolvedAnimation } as Theme,
    shortcuts: _shortcuts,
  }
}
