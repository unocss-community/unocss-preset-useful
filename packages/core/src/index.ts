import type { UserConfig } from 'unocss'
import type { UsefulOptions, UsefulTheme } from './types'
import { definePreset, mergeConfigs } from '@unocss/core'
import { shortcuts as builtInShortcuts, extractors, postprocess, preflights, rules, variants } from './core'
import { PRESET_NAME } from './meta'
import { resolveOptions } from './resolve'

export * from './utils'

export type { UsefulOptions, UsefulTheme }

export const presetUseful = definePreset<UsefulOptions, UsefulTheme>(async (options) => {
  const resolvedOptions = await resolveOptions(options ?? {})
  const { theme, meta, shortcuts: userShortcuts } = resolvedOptions
  const shortcuts = (userShortcuts === true || (typeof userShortcuts === 'object' && userShortcuts.default !== false))
    ? builtInShortcuts
    : []

  return {
    name: `unocss-preset-${PRESET_NAME}`,
    layers: {
      [PRESET_NAME]: 2,
    },
    rules,
    theme,
    variants: variants(resolvedOptions),
    shortcuts,
    extractors,
    postprocess: postprocess(resolvedOptions),
    presets: meta.presets,
    transformers: meta.transformers,
    preflights: preflights(resolvedOptions),
    options: resolvedOptions,
  }
})

export default presetUseful

export function defineConfig<T extends object = UsefulTheme>(config: UserConfig<T>) {
  return config
}

export function defineUsefulConfig<T extends object = UsefulTheme>(options: UsefulOptions = {}, config: UserConfig<T> = {}) {
  return mergeConfigs([
    defineConfig<T>({
      presets: [
        presetUseful(options) as any,
      ],
    }),
    config,
  ])
}
