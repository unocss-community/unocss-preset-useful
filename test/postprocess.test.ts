import { createGenerator, presetUno } from 'unocss'
import { describe, expect, test } from 'vitest'
import type { UsefulOptions } from '../src'
import { presetUseful } from '../src'

function generateUno(options: UsefulOptions = {}) {
  return createGenerator({
    presets: [
      presetUno(),
      presetUseful(options),
    ],
  })
}

describe('postprocess with unColor', () => {
  const code = 'bg-red text-blue'

  test('base', async () => {
    const uno = generateUno({
      unColor: true,
    })

    const { css } = await uno.generate(code, { preflights: false })

    expect(css).toMatchSnapshot()
  })

  test('with string', async () => {
    const uno = generateUno({
      unColor: '--test-color',
    })

    const { css } = await uno.generate(code, { preflights: false })

    expect(css).toMatchSnapshot()
  })
})
