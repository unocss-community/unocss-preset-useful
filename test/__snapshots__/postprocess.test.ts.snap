// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`presetUseful postprocess with unColor > base 1`] = `
"/* layer: default */
.bg-red{--un-bg-opacity:1;background-color:rgb(248 113 113 / var(--un-bg-opacity));}
.text-blue{--un-text-opacity:1;color:rgb(96 165 250 / var(--un-text-opacity));}"
`;

exports[`presetUseful postprocess with unColor > with any string 1`] = `
"/* layer: default */
.bg-red{--un-bg-opacity:1;background-color:rgb(248 113 113 / var(--un-bg-opacity));}
.text-blue{--un-text-opacity:1;color:rgb(96 165 250 / var(--un-text-opacity));}"
`;
